File: SparkCourses/22-02-2019/Section3/most-popular-superhero.py
from pyspark import  SparkContext, SparkConf

conf = SparkConf().setMaster("local").setAppName("MostPopularSuperhero")
sc = SparkContext(conf=conf)

def parseLine(lines):
    fields = lines.split("\"");
    return (int(fields[0]), fields[1])
def CountCoOccurence(lines):
    fields = lines.split()
    return (int(fields[0]), len(fields)-1)
names = sc.textFile("/home/bjit-557/BigData/February/SparkCourses/22-02-2019/Section3/Marvel-Names.txt")
namesRdd = names.map(parseLine)
lines = sc.textFile("/home/bjit-557/BigData/February/SparkCourses/22-02-2019/Section3/Marvel-Graph.txt")
pairings = lines.map(CountCoOccurence)
pairings.foreach(lambda x:print(x))
totalFriendsByHero = pairings.reduceByKey(lambda x,y:x+y)
totalFriendsByHero.foreach(lambda x:print(x))
flipped = totalFriendsByHero.map(lambda x:(x[1],x[0]))
flipped.foreach(lambda x:print(x))
mostPopular = flipped.max()
print(mostPopular)

mostPopularName = namesRdd.lookup(mostPopular[1])[0]
print(mostPopularName + " is most popular hero with " + str(mostPopular[0]) + " CoOccurancs")

File: scala/most_popular_superhero.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf

object most_popular_superhero {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local").setAppName("MostPopularSuperHero")
    val sc = new SparkContext(conf)
    val srcFile = "../SparkCourses_Scala/data/Marvel-Names.txt"
    val names = sc.textFile(srcFile)
    val namesRdd = names.map(map_ParseLine)


    val lines = sc.textFile("../SparkCourses_Scala/data/Marvel-Graph.txt")
    val pairings = lines.map(CountCoOccurances)
    val totalFriendsByHero = pairings.reduceByKey(reducer)
    val flipped = totalFriendsByHero.map(swap)
    val mostPopular = flipped.max()
    val mostPopularNames = namesRdd.lookup(mostPopular._2)(0)
    println(mostPopularNames)
  }

  def map_ParseLine(lines:String) : (Int, String) = {
    val fields = lines.split(" ");
    var name="";
    var i=1;
    while(i<fields.length)
      {
        name=name+" "+fields(i);
        i=i+1;
      }
    return  (fields(0).trim().toInt, name.replace("\"",""))
  }

  def CountCoOccurances(x:String) : (Int, Int) = {
    val fields = x.split(" ")
    return (fields(0).toInt, fields.length-1)
  }

  def reducer(x:Int, y:Int):(Int) = {
    return x+y
  }
  def swap(x:(Int, Int)) : (Int, Int) = {
    return (x._2, x._1)
  }
}
